network/p2p/dns/cache.go

```diff
@@ -0,0 +1,88 @@
+package dns
+
+import (
+	"net"
+	"sync"
+	"time"
+)
+
+// defaultTimeToLive is the default duration a dns result is cached.
+const defaultTimeToLive = 5 * time.Minute
+
+// cache is a ttl-based cache for dns entries
+type cache struct {
+	sync.Mutex
+	ttl      time.Duration // time-to-live for cache entry
```

Comment:
Agree with your elaboration. But as you also mentioned, the DNS resolver of libp2p does not provide any latency insight. So are you suggesting to have a long default TTL? e.g., an hour? or 24?

Reply:
@yhassanzadeh13 Thanks for replying to the comment! :-)

I'm suggesting that there are two issues to be considered:

1. That the usual way to design DNS caches is to implement the TTL as individual to each unique domain, e.g.:

"Caching name servers (DNS caches) store DNS query results for a period of time determined in the configuration (time-to-live) of each domain-name record." [1]

If we design it differently -- i.e. having a global TTL for all domains, instead of respecting the individual domain TTLs -- then this is confusing to all others who are used to working with DNS, including those people who specifically create the DNS records expecting DNS caches to respect the TTL they set in the DNS record for a particular domain.

> the DNS resolver of libp2p does not provide any latency insight

So are we okay with using a non-standard DNS caching strategy just because the DNS resolver is not supplying enough info? Options which spring to mind would be to enhance the libp2p resolver to return the extra info, or to use a different resolver?

The reason I'm bringing it up here is not to force us down this path, but to at least to make it a conscious and debated design decision :-)

2. That we can predict the effectiveness of the DNS cache before testing out this implementation. And my prediction is that in its current form, using the global TTL of 5 minutes for all domains, and using the passive timeout mechanism, and this connection purging happening every 10 minutes, then I'm sticking my neck out and predicting many cache misses on a regular basis :-) And I elaborated on this further in this comment [2].

[1] https://en.wikipedia.org/wiki/Name_server#Caching_name_server
[2] https://github.com/onflow/flow-go/pull/1086#discussion_r695149618